### ISSUE TITLE
Investigate build reattaching through worker restarts

### ISSUE THREAD

**Issue by clarafu** [concourse/concourse#3310]
The ATC is supposed to re-attach and retry requests to workers if they go away while a build is in-flight. It looks like there may be some cases where that doesn't happen.

One case where we know there's a gap is when re-attaching to resource actions (that's the `unexpected EOF` error you may see in `put` and `get` steps). This happens when a resource call (`/in`, `/out`) completes while the ATC wasn't attached, and we missed the response. This will be fixed with concourse/rfcs#1.

But aside from that, there have been a few failures observed in builds of `testflight` and builds that were running while we deployed our own Concourse.

For example:

* https://ci.concourse-ci.org/teams/main/pipelines/concourse/jobs/bosh-topgun/builds/257
* https://ci.concourse-ci.org/teams/main/pipelines/concourse/jobs/bosh-prod-deploy/builds/117

Let's test this out by hand and see if we can get things to fail when workers are restarted willy-nilly.

- @vito & @clarafu

**Comment by vito** [comment]
@kcmannem @ddadlani here's the issue I mentioned in retro - tossing it in Runtime somewhere!

**Comment by ddadlani** [comment]
Related to #2581

**Comment by vito** [comment]
Happened again:

* https://ci.concourse-ci.org/teams/main/pipelines/concourse/jobs/bosh-topgun/builds/279

**Comment by ddadlani** [comment]
We think we've found the explanation for this. Garden writes an exit code for the process to a file under the appropriate depot dir. When it completes waiting for the process, it'll open and read the file to determine the error condition.

https://github.com/cloudfoundry/guardian/blob/4722aa257d136051bebe6df22f8f1cdce73fafb1/rundmc/execrunner/dadoo/execrunner.go#L379-L381

We can see that the OS.Stat fails if the file didn't get written on time. This results in a error like: `connection: process error: could not find the exitcode file for the process: stat /worker-state/depot/1329bc8a-6937-4e62-50c7-4d70a9e8846c/processes/task/exitcode: no such file or directory`
In other situations we've seen an `Unexpected EOF` error be raised instead. This is returned by the `Open` syscall.

We will investigate further but it's likely that we just need to retry the step if this occurs.

**Comment by vito** [comment]
@ddadlani I've never seen that 'could not find exitcode' error before. How is Garden being restarted?

FWIW, the failing test in TopGun is running a process that definitely should not have exited while Garden was down - it loops until someone creates `/tmp/stop-waiting` in the container.

I had a hunch that `unexpected EOF` was actually a network error, since usually Garden annotates any error messages that it actually returns (just like the other error). Is that being propagated from Garden or is it only visible on the ATC side? I thought it might be happening when the ATC `.Wait` call gets interrupted from the other side going away, and right now `retryhttp` doesn't retry on `unexpected EOF` so it ends up being fatal.

I'd like to avoid retrying the step if this turns out to just be a network error that we need to catch in the connection retry logic.

**Comment by ddadlani** [comment]
Based on conversation in IPM: 

We should look at adding the `unexpected EOF` error into the `retryhttp` logic.

This could happen because the TSA shuts down the forwarded addresses (which would probably occur when the worker goes away) while still accepting connections.

**Comment by topherbullock** [comment]
@kcmannem and I traced down the path of how we're creating and using Garden Connections.
We laughed, we cried (well.. I cried a bit from laughing), and we wrote this PR to lay out some of the connection building in a way that is a little cleaner #3478

We _think_ that we can just add the EOF error to the `retryhttp` logic here:  https://github.com/concourse/retryhttp/blob/master/retryer.go#L35-L44
For reference, Go's default client sets the `accept-encoding` header to `gzip`. When you drop a gzipped response midway, the footer is missing. Which causes the gzip reader to error out with `Unexpected EOF`.

but we also observed some cases where it seems like we could have connections retrying for a long while ( eg. when we killed the garden server in the docker-composed worker ), and want to verify where this is blocking (funnily enough delve didn't help here because setting and stopping at breakpoints meant Garden already sent some data.. )
